proposals/2025-02-24_secret_providers.md

```diff
@@ -0,0 +1,182 @@
+
+This proposal introduces secret discovery, similar to service discovery, where different secret providers can contribute code to read secrets from their respective APIs. This would allow for more secure and dynamic secret retrieval, eliminating the need to store secrets in the filesystem and reducing the potential for unauthorized access.
+
+### Pitfalls of the current solution
+
+Storing secrets in the filesystem poses risks, especially in environments like Kubernetes, where any pod on a node can access files mounted on that node. This could expose secrets to attackers. Additionally, configuring secrets through the filesystem often requires extra setup steps in some enviroments, which can be cumbersome for users.
+
+Storing secrets inline can also pose risks, as the configuration file may still be accessible through the filesystem. Additionally it can lead to configuration files becoming cluttered and difficult to manage.
+
+## Goals
+
+Goals and use cases for the solution as proposed in [How](#how):
+
+* Allow Prometheus to read secrets remotely from secret providers.
+* Introduce secret discovery, similar to service discovery, where different secret providers can contribute code to read secrets from their respective API.
+
+### Audience
+
+* Prometheus maintainers
+* Alertmanager maintainers
+* Secret providers interested in contributing code
+* Users looking to use secret providers
+
+## Non-Goals
+
+* Implement a variety of secret providers.
+
+## How
+
+
+### Interfaces
+
+Secret providers will be created from their configurations through the following interface:
+
+```
+type SecretProviderConfiguration interface {
+	// Returns the secret provider for the given configuration.
+	Load() (SecretProvider, error)
+}
+```
+
+Secret providers will have to satisfy the following interface. Secret providers will be expected to be long lived to allow for caching. However they will be re-instantiated in case of a configuration change. The Fetch method will be called before each http request done through `prometheus/common/config`
+
+```
+type SecretProvider interface {
+	// Returns the secret value for the given configuration.
+  Fetch(ctx context.Context, secretRef string) (string, error)
+}
+```
+
+### Configuration
+
+Globally there will be a section to configure secrete providers to be used throught the config file. Here is an example:
+
+```
+global:
+  secret_providers:
+  - name: my_secret_provider
+    kubernetes_sp_config:
+      namespace: ns1
+```
+
+
+For secret related fields under http_config, a new `ref` variant will be added that can reference these secret providers. For instance, basic_auth will have the following form
+
+```
+basic_auth:
+...
+  password: <secret>
+  password_file: <string>
+  password_ref:
+    <string>: <string>
```

Comment:
What's the format of key and value here?

Reply:
Both issues should be mentioned in the new draft